kafka-client/src/main/java/dev/responsive/db/CassandraClient.java

```diff
@@ -348,37 +293,43 @@ public void prepareStatements(final String tableName) {
             .build()
     ));
 
-    revokePermit.computeIfAbsent(tableName, k -> session.prepare(
+    getMetadata.computeIfAbsent(tableName, k -> session.prepare(
+        QueryBuilder
+            .selectFrom(tableName)
+            .column(EPOCH.column())
+            .column(OFFSET.column())
+            .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .build()
+    ));
+
+    setOffset.computeIfAbsent(tableName, k -> session.prepare(
         QueryBuilder
             .update(tableName)
             .setColumn(OFFSET.column(), bindMarker(OFFSET.bind()))
-            .setColumn(PERMIT.column(), QueryBuilder.literal(UNSET_PERMIT))
             .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
-            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(OFFSET_KEY)))
-            .ifColumn(OFFSET.column()).isLessThan(bindMarker(OFFSET.bind()))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .ifColumn(EPOCH.column()).isEqualTo(bindMarker(EPOCH.bind()))
             .build()
     ));
 
-    acquirePermit.computeIfAbsent(tableName, k -> session.prepare(
+    reserveEpoch.computeIfAbsent(tableName, k -> session.prepare(
         QueryBuilder
             .update(tableName)
-            .setColumn(PERMIT.column(), bindMarker("np"))
+            .setColumn(EPOCH.column(), bindMarker(EPOCH.bind()))
             .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
-            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(OFFSET_KEY)))
-            .ifColumn(OFFSET.column()).isLessThan(bindMarker(OFFSET.bind()))
-            .ifColumn(PERMIT.column()).isEqualTo(bindMarker("op"))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .ifColumn(EPOCH.column()).isLessThan(bindMarker(EPOCH.bind()))
             .build()
     ));
 
-    finalizeTxn.computeIfAbsent(tableName, k -> session.prepare(
+    ensureEpoch.computeIfAbsent(tableName, k -> session.prepare(
         QueryBuilder
             .update(tableName)
-            .setColumn(PERMIT.column(), QueryBuilder.literal(UNSET_PERMIT))
-            .setColumn(OFFSET.column(), bindMarker(OFFSET.bind()))
+            .setColumn(EPOCH.column(), bindMarker(EPOCH.bind()))
```

Comment:
this makes me a little bit nervous, because the database could technically optimize this whole statement away and still be "correct". I wonder if it would make sense to set another column with some random uuid that always changes or something to ensure this always is applied.

Reply:
I'm not too concerned as the guarantee is "Use the IF keyword to introduce a condition that must return TRUE for the update to succeed." - `IF` and `WHERE` are not equivalent, I can imagine it optimizing it away with a `WHERE` clause.

OTOH I don't think there's much overhead for adding an additional column, but I'll research that tomorrow as a follow up.

Reply:
ah - I did not catch that distinction. Very cool